src/haven.rs

```diff
@@ -114,7 +116,67 @@ pub async fn haven_loop(ctx: DaemonContext, haven_cfg: HavenForwardConfig) -> an
             from_dock: _,
             to_port: _,
         } => tcp_forward(ctx, haven_cfg).await,
-        ForwardHandler::SimpleProxy { listen_dock: _ } => todo!(),
+        ForwardHandler::SimpleProxy { listen_dock: _ } => simple_proxy(ctx, haven_cfg).await,
+    }
+}
+
+async fn simple_proxy(
+    ctx: DaemonContext,
+    haven_cfg: HavenForwardConfig,
+) -> Result<(), anyhow::Error> {
+    let listen_dock = match haven_cfg.handler {
+        ForwardHandler::SimpleProxy { listen_dock } => listen_dock,
+        _ => anyhow::bail!("invalid config for simple_proxy"),
+    };
+
+    let haven_id = IdentitySecret::from_bytes(&earendil_crypt::kdf_from_human(
+        &haven_cfg.identity_seed,
+        "identity_kdf_salt",
+    ));
+    log::info!(
+        "simple proxy haven fingerprint: {}",
+        haven_id.public().fingerprint()
+    );
+
+    let earendil_skt = Socket::bind_haven_internal(
+        ctx.clone(),
+        haven_id,
+        Some(listen_dock),
+        Some(haven_cfg.rendezvous),
+    );
+
+    let mut listener = StreamListener::listen(earendil_skt);
+
+    let reaper = TaskReaper::new();
+    loop {
+        let mut earendil_stream = listener.accept().await?;
+
+        log::debug!("accepted simple proxy forward");
+        reaper.attach(smolscale::spawn(async move {
+            // the first 2 bytes of the stream encode the byte-length of the subsequent `hostname:port`
+            let mut len_buf = [0; 2];
+            earendil_stream.read_exact(&mut len_buf).await?;
+            let len: u16 = u16::from_be_bytes(len_buf);
+
+            let mut hostname_buf = vec![0; len as usize];
+            earendil_stream.read_exact(&mut hostname_buf).await?;
+
+            let addr = String::from_utf8_lossy(&hostname_buf).into_owned();
+            log::error!("address is {addr}");
+            let mut addrs = smol::net::resolve(addr.clone()).await?;
```

Comment:
I looked at the docs again, and you can actually pass a string `hostname:port` directly to TcpStream::connect, sidestepping all of this